lib/trace/tchannel.js

```diff
@@ -0,0 +1,163 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var _ = require('underscore');
+var util = require('util');
+
+var errors = require('../errors');
+
+var core = require('./core');
+
+//
+// TChannel forwarder
+//
+// opts = {
+//     sink: {
+//         // REQUIRED
+//         "type": "tchannel"
+//         "host-port": '1.2.3.4:5000',
+//
+//         // OPTIONAL/HAS DEFAULTS
+//         "service-name": 'ringpop',
+//         endpoint: 'trace-<event>'
+//     }
+// }
+//
+
+var tchannelOptsDefaults = {
+    sink: {
+        type: 'tchannel',
+        serviceName: 'ringpop',
+        endpoint: 'trace'
+    }
+};
+
+function TChannelTracer(ringpop, config, opts) {
+    opts.sink = _.extend(
+        {},
+        tchannelOptsDefaults.sink,
+        config.sinkOptsDefaults.tchannel,
+        opts.sink);
+    core.Tracer.call(this, ringpop, config, opts);
+}
+util.inherits(TChannelTracer, core.Tracer);
+
+
+TChannelTracer.prototype.validate = function validate() {
+    var sinkOpts = this.opts.sink;
+    return (
+        core.Tracer.prototype.validate.call(this) &&
+        sinkOpts &&
+        sinkOpts.type === 'tchannel' &&
+        sinkOpts.serviceName &&
+        sinkOpts.endpoint
+    );
+};
+
+TChannelTracer.prototype.connectSink = function connectSink(callback) {
+    if (callback) {
+        this.once('connectSink', callback);
+    }
+
+    var sinkOpts = this.opts.sink;
+    this.reqOpts = _.extend(
+        {
+            host: sinkOpts.hostPort,
+            timeout: 1000,
+            serviceName: 'ringpop',
+            hasNoParent: true,
+            trace: false,
+            retryLimit: 3,
+            headers: {
+                'as': 'raw',
+                'cn': 'ringpop'
+            }
+        },
+        _.omit(sinkOpts, 'type', 'hostPort')
+    );
+
+    // trigger early identification, minimize latency
+    this.ringpop.channel.waitForIdentified(
+        _.pick(this.reqOpts, 'host'),
+        this._remoteError.bind(this, 'connectSink'));
+};
+
+TChannelTracer.prototype.disconnectSink = function disconnectSink(callback) {
+    var self = this;
+    if (callback) {
+        this.once('disconnectSink', callback);
+    }
+
+    process.nextTick(function() {
+        self.disconnectSource();
+        self.emit('disconnectSink');   
+    });
+};
+
+TChannelTracer.prototype.send = function send(blob, callback) {
```

Comment:
`send()` acts as the event listener and 'event forwarder'. Are you assuming here that `blob` is the only argument passed to event listeners? Or am I missing something?